### ISSUE TITLE
Portal Use Logging

### ISSUE THREAD

**Issue by kittyclaus** [stargate-rewritten/Stargate-Bukkit#266]
I want some way to log who is entering, exiting, building, and destroying portals. I gave my server members free rein on personal portals, and we have a town nexus. It'd just be nice to check in and see what happened when there is conflict.

**Comment by Thorinwasher** [comment]
This would be perfect for an addon

**Comment by Pheotis** [comment]
> This would be perfect for an addon

I'm not convinced that this needs to be an addon.
Could possibly have an enableLog toggle in the config that outputs to a stargate.log file.

But yeah, something beyond that (for example, per-portal logging) would be ideal for stargate-interfaces.
A possible alternative might be expanding on #263 and passing even more stargate information to coreprotect.

**Comment by Thorinwasher** [comment]
#263 is not really related to this. I just thought it would be amazingly easy to implement to an addon, that is just a event being cancelled, which shouldn't.

I'm thinking having something that stores every action one stargate has experienced, then similar to coreprotect, you can search for data on said portal. This would be relatively easy to implement. Any good names for it?

**Comment by Pheotis** [comment]
> #263 is not really related to this. I just thought it would be amazingly easy to implement to an addon, that is just a event being cancelled, which shouldn't.
We cancel the button event for good reason; it is how we allow stargates to work regardless of region protection without allowing other plugins to send toggle permission errors.
Since we cancel the event for button permission on stargates, fixing #263 will probably involve using coreprotect's api to inform it directly of the event.

If we are going to need to use the api anyways, may as well make full use of it to improve our integration with coreprotect on the whole.

> I'm thinking having something that stores every action one stargate has experienced, then similar to coreprotect, you can search for data on said portal. This would be relatively easy to implement. Any good names for it?

I think this is two systems:
__Simple System -- a logger within the base plugin__
- A `logPortals` option in the config that by default is false
- If true, a `stargate.log` file that simply gets appended with player, origin, and destination for all teleport events.

__Advanced System -- basically what you described, and used in an addon (probably sg-interfaces)__
Assuming this is part of sg-interfaces, I think something like `/sgc audit` would probably work best.

**Comment by Thorinwasher** [comment]
> We cancel the button event for good reason; it is how we allow stargates to work regardless of region protection without allowing other plugins to send toggle permission errors.

Good point, but then this would be covered in the addon I mentioned in #263

**Comment by Pheotis** [comment]
> > We cancel the button event for good reason; it is how we allow stargates to work regardless of region protection without allowing other plugins to send toggle permission errors.
> Good point, but then this would be covered in the addon I mentioned

It does make sense to cover this in an addon, but if we do that, I still think using coreprotect's api is still the best idea.

Bypassing third-party plugin switch protection to allow access to stargates is a key feature of stargate, and is what makes stargate's access nodes as useful as they are. It has been a key feature of stargate since Dinnerbone's implementation. 
If we allow this protection to be disabled, it should probably be as a hidden config option or as an addon config option (and should be false by default).

Since this will involve coreprotect's api, we should probably pass as much to coreprotect as we can (even if its in an addon).
I personally think sg-mechanics or possibly a theoretical catch-all sg-compatibility would be an ideal addon target for this.

> I really think we should avoid making the plugin larger than it already is. That's why I don't want to add much more functionality to it
While I agree that the core functionality is something for addons, I think a base log file fits well within base.
Therefore, I strongly disagree for the following reasons:

- Given that this would be a file-based, admin-facing, feature, it is compatible with the dinnerbone philosophy.
- We have a debug folder, and adding a stargate.log file wouldn't cause much clutter.
- We have a *console* logging section of the config; adding a simple *access* log would be quite easy and wouldn't add much bloat.
- Basically every other plugin does this natively; having some minimal form of auditing in base seems worthwhile.
- This is exceptionally easy to implement -- it's simply appending a text file with basic teleport events.
- Even if one feels that this is bloated, given that it is ridiculously simple to implement, I think base .log capacity could at the very least be a hidden config option.

**Comment by Thorinwasher** [comment]
> This is exceptionally easy to implement -- it's simply appending a text file with basic teleport events.

Sure is, but this style of thought is what has delayed the full plugin release for 1 and a half years. What makes something bloated is not just one thing, but the `many` things. If this could be covered in an addon with so much better style, why not just have it there instead?

**Comment by Pheotis** [comment]
> > This is exceptionally easy to implement -- it's simply appending a text file with basic teleport events.
> 
> Sure is, but this style of thought is what has delayed the full plugin release for 1 and a half years. What makes something bloated is not just one thing, but the `many` things. If this could be covered in an addon with so much better style, why not just have it there instead?

### Benefits of the main plugin from the perspective of users:
1. Easily accessible to all users; especially including those using ESR releases and those from non-english communities.
2. Complete compliance with the dinnerbone philosophy, and thus, complete immersion in-game.
3. Support: main is our number one priority when it comes to maintenance
4. Easy of understanding; the base branch should have everything one needs, with nothing one does not.

The goal of the main branch is to package *just enough* features in the most intuitive way possible.
**With the main branch, deployments should have enough features to accomplish all reasonable use cases.**
Therefore, the main branch should be usable by most servers in its own right.

An average server should never feel the *need* to get an addon; it should just that -- an expansion upon existing content; i.e., a "Nice-to_have".

### Benefits of addons for perspective users.
1. Comprehensive and tailored to their whims.
2. Expands upon the base functionality in novel ways and adds excitement to the project.

The goal of addons is to allow cool functionality to exist, even when it does not make sense in the base plugin.
In this case, "Doesn't make sense" can be lack of intuitiveness, lack of immersion (dinnerbone philosophy), or lack of applicability towards an average use case.

Addon features should expand upon the needs of average use cases, but should never be the sole implementation of average users.

## Thus
Adding extra flags is an extra because flags exist in the main version; there is a way for users to network their portals in interesting ways, and sg-mechanics simply expands upon them.

Adding extra customisations is an extra because some level of customisations (i.e. dye colours and basic sign formats) exists in the base plugin, and sg-customizations simply expands upon them.

Adding extra interfaces is an extra because there already exist a level of interfaces in the game (sign behaviour, config files, etc), and sg-interfaces simply expands upon them.

## therefore
There are currently no reasonable ways (short of tailing and filtering console logs) for admins to tell portal usage;
There is currently no way to audit stargates for potential misuse.

Is this an extra? Is it not something needed by an average use case?
I strongly feel that it is not; I feel that this is something that is useful for every server.
Installing an addon to get basic logs is overkill.

Does that mean the base plugin needs to log all events and needs any form of an inspector? I don't think so.
But I do think it means that a `stargate.log` file at a minimum should be included in the main plugin.

**Comment by Thorinwasher** [comment]
> Does it lack intuitiveness? I don't feel it does. Does it break immersion (i.e. philosophy)? I don't feel it does.

I agree, its not unintuitive, and of course does not break immersion, as would be an admin tool.
> And most notably, is this an extra; is it not something not needed within an average use case? I strongly feel that it is not; that it is a core mechanic and not an extra; I feel that this is something that is useful for every server. That is to say, I feel that installing an addon to get basic logs is overkill.

Here comes the issue with just having basic logs: they will spam the fuck out of the console, so to get some actual information on something, you would still need an actual interface like coreprotect provides. I don't see the use case for just spamming a file with messages.

That's why I'm not interested enough in implementing it.

**Comment by Pheotis** [comment]
> > And most notably, is this an extra; is it not something not needed within an average use case? I strongly feel that it is not; that it is a core mechanic and not an extra; I feel that this is something that is useful for every server. That is to say, I feel that installing an addon to get basic logs is overkill.
> 
> Here comes the issue with just having basic logs: they will spam the fuck out of the console, so to get some actual information on something, you would still need an actual interface like coreprotect provides. I don't see the use case for just spamming a file with messages.

> *[paraphrased] it will spam the fuck out of the console*

In terms of spamming console, I do not think that would be an issue;
By definition, the output to a `stargate.log` file would not show up in console/latest.log, unless the user had their logger set to `finer`, in which case they explicitly opted for any such spam.

> *[paraphrased] I don't see the use case for just spamming a file with messages [since] you would still need an actual interface*

I think this discussion boils down to that; at least for smaller servers, I feel that there is a significant use case for such a file.
I feel that there is a reason why a great many major projects (including essx, chestshop, etc.) still offer similar file-based, simplified, logs despite the existence of better alternatives.

Assuming a stargate.log implementation is limited to appending the bare minimum information (portal entered, destination, player, timestamp, and possibly cost), such a solution presents significant utility despite its lack of detail.


To first address the obvious counterpoint, yes, a `stargate.log` file is as obnoxious as it is useless for large servers.
Any server with over 10 ACP would probably never use it, and instead would opt for an interface (i.e. an addon and/or co) or a script of some sort (i.e. interpreting the main logs directly). If servers of this sort represented most of SG's traditional userbase, I would agree that this feature had no place in main.

Yet, based on legacy's peak metrics, large servers have traditionally made up less than 30% of SG's net userbase.
It is on the smaller servers (those expected to make up the majority of use cases) that this feature shines:

- It is difficult to overstate the simplicity and intuitiveness of a basic log file
  - The text log format has been around for decades and is very well understood.
  - Log files require no dependencies, very minimal understandings of english, and no technical knowledge.
- On small servers, such a log file would likely get <20 entries appended per day, and thus, remains human readable/interpretative (with the additional option of using ctrl+f)
  - This reason would be totally overkill to install a full add-on just to know basic info about portal usage.
- Despite the small size and simplicity, one can nonetheless find answers to key questions such as:
  - Has anyone been using the plugin on my server and/or is it working?
  - Are particular gates/areas used more than others? What gates are preferred by certain players?
  - Has anyone been using gates in ways they are not supposed to?


At the very least, given the negligible implementation effort, I feel that the use cases presented above are strong enough to justify adding a hidden config option.

**Comment by Thorinwasher** [comment]
My main point is that I'm not interested in implementing this, more clearly why:
- It will delay the release. Oh but this is a small addition. Yes but it's one of _many_ additions, the amount of them will delay the project release indefinitely
- It's not something interesting to code. I'm doing this as a hobby, alright, got it?
- I don't like the concepts of it. It spams a file with a shit ton of messages which will be hard to get any data from, this might be what some user wants, but is it what they deserve?
- Stargate as a plugin has become really large, this is because of all of these small things. In essence the larger a plugin becomes, the probability of a bug to exist increases. You know the issues with that for a plugin that has not been released yet
